src/app/clusters/ecosystem-information-server/ecosystem-information-server.h

```diff
@@ -0,0 +1,211 @@
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+// This cluster is targeted by devices that are not resource constrained, for
+// that reason we use std containers to simplify implementation of the cluster.
+#include <map>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <app-common/zap-generated/cluster-objects.h>
+
+#include <app/AttributeAccessInterface.h>
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace EcosystemInformation {
+
+// This intentionally mirrors Structs::EcosystemDeviceStruct::Type but has ownership
+// of underlying types.
+class EcosystemDeviceStruct
+{
+public:
+    class Builder
+    {
+    public:
+        Builder(){};
+        Builder(const Builder &) = delete;
+
+        Builder & SetDeviceName(std::string aDeviceName, uint64_t aDeviceNameLastEditEpochUs);
+        Builder & SetBrigedEndpoint(EndpointId aBridgedEndpoint);
+        Builder & SetOriginalEndpoint(EndpointId aOriginalEndpoint);
+        Builder & AddDeviceType(Structs::DeviceTypeStruct::Type aDeviceType);
+        Builder & AddUniqueLocationId(std::string aUniqueLocationId, uint64_t aUniqueLocationIdsLastEditEpochUs);
+
+        // Upon success this object will have moved all ownership of underlying
+        // types to EcosystemDeviceStruct and should not be used afterwards.
+        std::unique_ptr<EcosystemDeviceStruct> Build();
+
+    private:
+        std::string mDeviceName;
+        uint64_t mDeviceNameLastEditEpochUs = 0;
+        EndpointId mBridgedEndpoint         = kInvalidEndpointId;
+        EndpointId mOriginalEndpoint        = kInvalidEndpointId;
+        std::vector<Structs::DeviceTypeStruct::Type> mDeviceTypes;
+        std::vector<std::string> mUniqueLocationIds;
+        uint64_t mUniqueLocationIdsLastEditEpochUs = 0;
+        bool mIsAlreadyBuilt                       = false;
+    };
+
+    EcosystemDeviceStruct(const EcosystemDeviceStruct &) = delete;
+
+    CHIP_ERROR Encode(const AttributeValueEncoder::ListEncodeHelper & aEncoder, const FabricIndex & aFabricIndex);
+
+private:
+    // Constructor is intentionally private. This is to ensure that it is only constructed with
+    // values that conform to the spec.
+    explicit EcosystemDeviceStruct(std::string aDeviceName, uint64_t aDeviceNameLastEditEpochUs, EndpointId aBridgedEndpoint,
```

Comment:
If this is built using a builder, would it make sense to use `&&` for the larger types (string, vectors)?

Reply:
100%, thanks!